### PR TITLE
kvserver: fix false positive in 'outstanding reproposal' assertion

### PR DESCRIPTION
**Pull request by tbg**
In #94633, I introduced an assertion that attempted to catch cases
in which we might otherwise accidentally end up applying a proposal
twice.

This assertion had a false positive, see the updated comment within.

I was able to reproduce the failure within ~minutes via
`./experiment.sh` in #97173 as of 33dcdef.

Better testing of these cases would be desirable. Unfortunately, while
there is an abstraction over command application (`apply.Task`), most
of the logic worth testing lives in `(*replicaAppBatch)` which is
essentially a `*Replica` with more moving parts attached. This does
not lend itself well to unit testing.

I had a run[^1][^2][^3] earlier this year to make log application
standalone[^4], but then didn't have enough time to follow through.
It would be desirable to do so at a later date, perhaps with
the explicit goals of having interactions like the one discussion
in this PR unit become testable.

No release note because unreleased (except perhaps in an alpha).

[^1]: #93239
[^2]: #93266
[^3]: #93309
[^4]: https://github.com/cockroachdb/cockroach/issues/75729

Closes #94633.

[^1]: https://github.com/cockroachdb/cockroach/pull/94633/files#diff-50e458584d176deae52b20a7c04461b3e4110795c8c9a307cf7ee6696ba6bc60R238

Epic: none
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 1]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
 			//
-			// If it was rejected with ProposalRejectionIllegalLeaseIndex, then the
-			// subsequent call to tryReproposeWithNewLeaseIndex[^1] must have returned an
-			// error (or the proposal would not be IsLocal() now). But that call
-			// cannot return an error for a proposal that is already superseded
-			// initially.
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:268]: Is `idx` the Raft log index? Should it be 101 here?
Or what does this duplicate mean? Remove it?

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 2]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+		if higherReproposalsExist := cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex); higherReproposalsExist &&
+			(!cmd.Rejected() || cmd.Rejection == kvserverbase.ProposalRejectionIllegalLeaseIndex) {
+			// If this entry was accepted, there must not be superseding reproposal,
+			// as nothing prevents it from applying as well in the common case.
+			//
+			// If this entry was rejected but only on account of the lease applied
+			// index, then the call to tryReproposeWithNewLeaseIndex[^1] must have
+			// returned an error (or the proposal would not be IsLocal() now; we
+			// unbind on success). But that call, by design, does not an error for a
+			// proposal that is already superseded.
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:256]: This paragraph focuses a bit on the "symptoms": what other methods return or not return. It's a bit hard to follow such logic because of the indirection. Is there a way to frame this in a stateless manner, maybe more abstractly - i.e. directly answer the question why a rejected MaxLeaseIndex can't be found superseded? And then, if needed, employ the additional argument on what the other methods return (which is more like a "proof").

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 3]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 101: new lease seq=2
+			//   idx 102: cmd X at LAI 10000, lease seq = 1
+			//   idx 103: cmd X at LAI 20000, lease seq = 1
+			//
+			// When we apply index 102, we will see a permanent rejection but the
+			// proposal is local (since we don't unlink it if it is already superseded
+			// initially, see `retrieveLocalProposals`) and superseded (by idx 103). A
+			// permanent rejection dooms all reproposals (including the superseding
+			// one) to the same fate, so idx 103 will get rejected just like idx 102.
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:285]:
```suggestion
			// one) to the same fate, so idx 103 will get rejected just like idx 102
```

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 4]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//
+			// Command `X` at log position 100 will get reproposed with a new lease
+			// index (as idx 99 consumed the LAI). But now the lease might also change
+			// hands, and an identical reproposal of index 100 might be appended, too:
+			//
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 101: new lease seq=2
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:277]: Where has the duplicate `idx 100` gone?

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 5]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			//   idx 102: cmd X at LAI 10000, lease seq = 1
+			//   idx 103: cmd X at LAI 20000, lease seq = 1
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:279]: Why 2 reproposals here? Is this because there was one before and one after the lease change? So the duplicate `idx 100` above is an in-flight `idx <TBD>` reproposal?

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 6]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//
+			// Command `X` at log position 100 will get reproposed with a new lease
+			// index (as idx 99 consumed the LAI). But now the lease might also change
+			// hands, and an identical reproposal of index 100 might be appended, too:
+			//
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:276]: Maybe add MaxLeaseIndex (MLI?) to the entries, to make it more obvious why things get rejected.

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 7]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			//   idx 101: new lease seq=2
+			//   idx 102: cmd X at LAI 10000, lease seq = 1
+			//   idx 103: cmd X at LAI 20000, lease seq = 1
+			//
+			// When we apply index 102, we will see a permanent rejection but the
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:281]: Why is this rejection permanent? Is it something we assume, or it should be obvious from this example?

What is an example of a permanent (non-MaxLeaseIndex) rejection?

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 8]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			// If this entry was accepted, there must not be superseding reproposal,
+			// as nothing prevents it from applying as well in the common case.
+			//
+			// If this entry was rejected but only on account of the lease applied
+			// index, then the call to tryReproposeWithNewLeaseIndex[^1] must have
+			// returned an error (or the proposal would not be IsLocal() now; we
+			// unbind on success). But that call, by design, does not an error for a
+			// proposal that is already superseded.
+			//
+			// If the command was rejected permanently, a superseding entry *can*
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:258]:
```suggestion
			// If the command was rejected permanently, a superseding proposal *can*
```

**Review by pav-kv — APPROVED** [review]
Flushing some comments, while I'm still digesting. Feel free to explain in text, or let's chat?

Proactively approving because this PR makes the check less strict and fixes the failures.

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 9]
```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
...
+			// For example, consider the following initial situation:
 			//
-			// If it was rejected with ProposalRejectionIllegalLeaseIndex, then the
-			// subsequent call to tryReproposeWithNewLeaseIndex[^1] must have returned an
-			// error (or the proposal would not be IsLocal() now). But that call
-			// cannot return an error for a proposal that is already superseded
-			// initially.
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_application_state_machine.go:267]: Should LAI be > 10000 here?

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_application_state_machine.go:267]: Nvm, this comment wasn't for this line. I intended to put this in line 278.

File: pkg/kv/kvserver/replica_application_state_machine.go [hunk 10]
```diff
@@ -150,6 +151,26 @@ func (sm *replicaStateMachine) NewBatch() apply.Batch {
 	return b
 }
 
+func formatReplicatedCmd(cmd *replicatedCmd) redact.RedactableString {
```
[the remark this answers is not in the document]
Comment (tbg) [reply pkg/kv/kvserver/replica_application_state_machine.go:154]: Yeah, this is sort of intermediate - I didn't want to get sidetracked writing a nice stringer and I didn't care about printing "everything". The standard way to implement `fmt.Stringer` is to implement `redact.SafeFormatter` and use that to implement the `String` method. But `replicatedCmd` is a pretty large nested struct so it'd be a lot of busy work. Generally a good idea to do it - just shied away from it for now.

File: pkg/kv/kvserver/replica_application_state_machine.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_application_state_machine.go:268]: Typo, it's 101

**Comment by pavelkalinnikov** [comment]
Flushed our pair programming edits.

File: pkg/kv/kvserver/replica_application_state_machine.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_application_state_machine.go:154]: SGTM. Fine with a TODO if you end up merging this.

**Comment by tbg** [comment]
Pavel and I pair-reviewed this PR and I think the assertion can still fire (erroneously). I'll update the PR with that scenario as well, and disable the assertion altogether. Will ping when done.

**Comment by tbg** [comment]
> Pavel and I pair-reviewed this PR and I think the assertion can still fire (erroneously).

I'm no longer sure what I was thinking here, I *think* the assertion is now mostly ok, except when introducing incorrect LAIs artificially. Either way, I'm removing it, rationale in https://github.com/cockroachdb/cockroach/pull/97347 which supersedes this PR.
